zkevm-circuits/src/keccak_circuit/util.rs

```diff
@@ -0,0 +1,320 @@
+            }
+            bytes.push(value);
+        }
+        bytes
+    }
+
+    pub(crate) fn value(bits: &[u8]) -> Vec<u8> {
+        debug_assert!(bits.len() % 8 == 0, "bits not a multiple of 8");
+
+        let mut bytes = Vec::new();
+        for byte_bits in bits.chunks(8) {
+            let mut value = 0u8;
+            for (idx, bit) in byte_bits.iter().enumerate() {
+                value += *bit << idx;
+            }
+            bytes.push(value);
+        }
+        bytes
+    }
+}
+
+/// Rotates a word that was split into parts to the right
+pub fn rotate<T>(parts: Vec<T>, count: usize, part_size: usize) -> Vec<T> {
+    let mut rotated_parts = parts;
+    rotated_parts.rotate_right(get_rotate_count(count, part_size));
+    rotated_parts
+}
+
+/// Rotates a word that was split into parts to the left
+pub fn rotate_rev<T>(parts: Vec<T>, count: usize, part_size: usize) -> Vec<T> {
+    let mut rotated_parts = parts;
+    rotated_parts.rotate_left(get_rotate_count(count, part_size));
+    rotated_parts
+}
+
+/// Encodes the data using rlc
+pub mod compose_rlc {
+    use eth_types::Field;
+    use halo2_proofs::plonk::Expression;
+
+    pub(crate) fn expr<F: Field>(expressions: &[Expression<F>], r: F) -> Expression<F> {
+        let mut rlc = expressions[0].clone();
+        let mut multiplier = r;
+        for expression in expressions[1..].iter() {
+            rlc = rlc + expression.clone() * multiplier;
+            multiplier *= r;
+        }
+        rlc
+    }
+}
+
+/// Scatters a value into a packed word constant
+pub mod scatter {
+    use super::BIT_SIZE;
+    use eth_types::Field;
+    use halo2_proofs::plonk::Expression;
+
+    pub(crate) fn expr<F: Field>(value: usize, count: usize) -> Expression<F> {
+        let mut packed = F::zero();
+        let mut factor = F::one();
+        for _ in 0..count {
+            packed += F::from(value as u64) * factor;
+            factor *= F::from(BIT_SIZE as u64);
+        }
+        Expression::Constant(packed)
+    }
+}
+
+/// The words that absorb data
+pub fn get_absorb_positions() -> Vec<(usize, usize)> {
+    let mut absorb_positions = Vec::new();
+    for j in 0..5 {
+        for i in 0..5 {
+            if i + j * 5 < 17 {
+                absorb_positions.push((i, j));
+            }
+        }
+    }
+    absorb_positions
+}
+
+/// Converts bytes into bits
+pub fn into_bits(bytes: &[u8]) -> Vec<u8> {
+    let num_bits = bytes.len() * 8;
+    let mut bits: Vec<u8> = vec![0; num_bits];
+
+    let mut counter = 0;
+    for byte in bytes {
+        for idx in 0u64..8 {
+            bits[counter] = (*byte >> idx) & 1;
+            counter += 1;
+        }
+    }
+
+    bits
+}
+
+/// Converts bits into bytes
+pub fn from_bits(bits: &[u8]) -> Word {
+    let mut value = Word::zero();
+    for (idx, bit) in bits.iter().enumerate() {
+        value += Word::from(*bit as u64) << idx;
+    }
+    value
+}
+
+/// Pack bits into a word
+pub fn pack(bits: &[u8]) -> Word {
```

Comment:
Should this be EVM word(256 bits) or Keccak word(64 bits)?

Reply:
I think everywhere I talk about words they are keccak words (because keccak doesn't directly have anything to do with EVM words I think it should be clear from context, or at least I hope so).

Reply:
The type `Word` here is `pub type Word = U256;` from `zkevm-circuits/eth_types`. Can we define a `type Word = u64` and replace U256 with it?

Reply:
Something >u64 is needed here because this will pack all 64 sparse bits into a single field element. With each sparse bit taking up 3 bits, the full packed word needs to be stored in at least `64*3 = 192` bits. Though I'm not sure anymore why I don't simply do all the math on a `Field` value directly, will see if it's easy to switch to that instead.